### PR TITLE
Test & infrastructure cleanup, stage 1

### LINKED ISSUE #831: Refactoring: IPC-JSON
The product itself and the test suite use two completely different implementations of IPC JSON protocol underlying DAP. This is partly because the product implementation is server-side-only, and the test implementation is client-side-only. Even so, it is mostly the same functionality, and there's no reason to implement it twice in such a radically different way. Furthermore, neither of those implementations has proper test coverage.

This item captures the work necessary to converge on a single implementation that is used throughout the product and the high-level test suite, with proper tests of its own.

### PR DESCRIPTION
**Pull request by int19h**
Add new JSON IPC implementation to be shared between the product and the tests.

Add pytest-based test support, and wire it up to setup.py and Travis.

Dial pylint down to complain about important things only.

Various minor fixes exposed by pylint.

Fixes #831.

### CODE REVIEW

**Comment by int19h** [comment]
The new IPC-JSON code is not wired to the product yet. The intent is to write functional tests utilizing it first, then use that to regression-test the product changes for migration.

Note that this change establishes a completely separate test suite under pytests/ - the intent is to fill it with functional tests for ptvsd proper in further stages, transferring (and cleaning up) code from the old tests, and eventually retire the old tests entirely.

To run the new suite:
```sh
pythonX.Y setup.py test
```
it will take care of dependencies etc.

**Comment by int19h** [comment]
Also, while reviewing, please give me feedback on the overall readability of the code and clarity of intent. The goal is to not have any code that is unnecessarily convoluted. Where it's not obvious what the code does even after careful reading, let me know so that I can add more comments, or refactor the code to be clearer.

For the tests especially, the goal is to have readability, even if it means extra verbosity.

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (karthiknadig) [inline .vscode/settings.json]: Ignore this file?

Comment (int19h) [reply .vscode/settings.json]: I removed it from gitignore on purpose - user-specific settings should be in user config, and this one is workspace-specific, and sets the appropriate knobs that everyone should be using (e.g. run flake8).

File: ptvsd/messaging.py [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+    def __init__(self, channel, seq):
+        self.channel = channel
+        self.seq = seq
+        self.response = None
+        self._got_response = threading.Event()
+        self._handler = lambda _: None
+
+    def _handle_response(self, success, command, error_message=None, body=None):
+        assert self.response is None
+        with self.channel._lock:
```
Comment (karthiknadig) [inline ptvsd/messaging.py]: Should `_lock` be a "public" property?

File: ptvsd/messaging.py [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+            success = message['success']
+            command = message['command']
+            error_message = message.get('message', None)
+            body = message.get('body', None)
+            self.on_response(seq, request_seq, success, command, error_message, body)
+        else:
+            raise IOError('Incoming message has invalid "type":\n%r' % message)
+
+    def on_request(self, seq, command, arguments):
+        handler_name = '%s_request' % command
```
Comment (karthiknadig) [inline ptvsd/messaging.py]: Low pri: Should handler decide the name?

Comment (int19h) [reply ptvsd/messaging.py]: You mean the suffix? I figured so long as there's a clear convention that guarantees no name clashes, it's fine. Just can't think of any practical scenario where the handler class would want to use custom names... but one possibility is to have a tiny handler wrapper that just has `request` and `event`, and then does its own dispatching to the real handler object by whatever rules it wants.

**Pull request merged**